# Knowledge Extraction



## Thực hiện trích xuất thực thể (Named Entity Recognition - NER) từ dữ liệu tài chính.

## Áp dụng mô hình hoặc quy trình xử lý để phân loại các thực thể thành các nhãn như: PER, ORG, LOC.

## Đánh giá và so sánh kết quả mô hình trên hai tập dữ liệu:

- flare_ner: Dữ liệu từ SEC filings.

- flare_finer_ord: Dữ liệu từ bài báo tài chính.

In [59]:
!pip install seqeval transformers datasets
import pandas as pd


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 2.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for seqeval: filename=seqeval-1.2.2-py3-none-any.whl size=16162 sha256=c330484daff1ecace7dba4a68819fa85a2e505f412a4ca78347e79bc31e6533c
  Stored in directory: /root/.cache/pip/wheels/bc/92/f0/243288f899c2eacdfa8c5f9aede4c71a9bad0ee26a01dc5ead
Successfully built seqeval


In [60]:
from google.colab import files
import pandas as pd

# Hiển thị cửa sổ chọn file
uploaded = files.upload()

# Giả sử bạn vừa upload 1 file CSV duy nhất, lấy tên file:
filename = list(uploaded.keys())[0]

# Đọc file CSV thành DataFrame
df = pd.read_csv(filename)

print(f"Đã tải lên file: {filename}")
print("Kích thước DataFrame:", df.shape)
df.head()

Saving flare_ner.csv to flare_ner.csv
Đã tải lên file: flare_ner.csv
Kích thước DataFrame: (408, 4)


,query,answer,label,text
0,In the sentences extracted from financial agre...,"SILICON VALLEY BANK, ORG\nBank, ORG\nCaliforni...","[O, O, O, O, O, O, O, O, O, O, O, O, B-ORG, I-...",This LOAN AND SECURITY AGREEMENT dated January...
1,In the sentences extracted from financial agre...,"Wellesley, LOC\nMassachusetts, LOC\nSilicon Va...","[O, O, B-LOC, O, B-LOC, O, O, O, O, O, B-LOC, ...","350 , Wellesley , Massachusetts 02481 doing bu..."
2,In the sentences extracted from financial agre...,"Bank, ORG","[O, O, O, O, O, O, O, O, B-ORG, O, O, O, O, O,...",This Agreement shall be construed to impart up...
3,In the sentences extracted from financial agre...,"Borrower, PER\nBank, ORG","[B-PER, O, O, B-ORG, O, O, O, O, O, O, O, O, O...",Borrower will pay Bank the unpaid principal am...
4,In the sentences extracted from financial agre...,"Bank, ORG\nBorrower, PER","[O, O, O, O, O, O, O, O, O, O, B-ORG, O, O, O,...",Subject to the terms and conditions of this Ag...


In [61]:
from google.colab import files
import pandas as pd

# Hiển thị cửa sổ chọn file
uploaded = files.upload()

# Giả sử bạn vừa upload 1 file CSV duy nhất, lấy tên file:
filename = list(uploaded.keys())[0]

# Đọc file CSV thành DataFrame
df = pd.read_csv(filename)

print(f"Đã tải lên file: {filename}")
print("Kích thước DataFrame:", df.shape)
df.head()

Saving flare_finer_ord.csv to flare_finer_ord.csv
Đã tải lên file: flare_finer_ord.csv
Kích thước DataFrame: (1075, 5)


,id,query,answer,label,token
0,finerord0,"In the list of tokens, identify 'Person', 'Loc...",H:O\nave:O\nyou:O\never:O\nfelt:O\nthat:O\nsud...,"[O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, ...","[H, ave, you, ever, felt, that, sudden, ,, int..."
1,finerord1,"In the list of tokens, identify 'Person', 'Loc...",That:O\nin:O\nless:O\nthan:O\nseven:O\nminutes...,"[O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, ...","[That, in, less, than, seven, minutes, ,, the,..."
2,finerord2,"In the list of tokens, identify 'Person', 'Loc...",And:O\nall:O\nbecause:O\nyou:O\nfailed:O\nto:O...,"[O, O, O, O, O, O, O, O]","[And, all, because, you, failed, to, prepare, !]"
3,finerord3,"In the list of tokens, identify 'Person', 'Loc...",You:O\nidiot:O\n!:O,"[O, O, O]","[You, idiot, !]"
4,finerord4,"In the list of tokens, identify 'Person', 'Loc...",And:O\nnow:O\nyou:O\n’re:O\non:O\na:O\ntrain:O...,"[O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, ...","[And, now, you, ’re, on, a, train, carriage, i..."


In [62]:
df_ner = pd.read_csv("flare_ner.csv")
df_finer = pd.read_csv("flare_finer_ord.csv")

# Hiển thị một vài dòng đầu
df_ner.head()
df_finer.head()


,id,query,answer,label,token
0,finerord0,"In the list of tokens, identify 'Person', 'Loc...",H:O\nave:O\nyou:O\never:O\nfelt:O\nthat:O\nsud...,"[O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, ...","[H, ave, you, ever, felt, that, sudden, ,, int..."
1,finerord1,"In the list of tokens, identify 'Person', 'Loc...",That:O\nin:O\nless:O\nthan:O\nseven:O\nminutes...,"[O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, ...","[That, in, less, than, seven, minutes, ,, the,..."
2,finerord2,"In the list of tokens, identify 'Person', 'Loc...",And:O\nall:O\nbecause:O\nyou:O\nfailed:O\nto:O...,"[O, O, O, O, O, O, O, O]","[And, all, because, you, failed, to, prepare, !]"
3,finerord3,"In the list of tokens, identify 'Person', 'Loc...",You:O\nidiot:O\n!:O,"[O, O, O]","[You, idiot, !]"
4,finerord4,"In the list of tokens, identify 'Person', 'Loc...",And:O\nnow:O\nyou:O\n’re:O\non:O\na:O\ntrain:O...,"[O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, ...","[And, now, you, ’re, on, a, train, carriage, i..."


# Biểu diễn dữ liệu đầu vào (Visualization or Inspection)

In [63]:
import random
print("Example from flare_ner:")
print(df_ner.iloc[random.randint(0, len(df_ner))])

print("Example from flare_finer_ord:")
print(df_finer.iloc[random.randint(0, len(df_finer))])


Example from flare_ner:
query     In the sentences extracted from financial agre...
answer                                        Borrower, PER
label     [O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, ...
text      " FUNDING DATE " is any date on which an Equip...
Name: 129, dtype: object
Example from flare_finer_ord:
id                                              finerord366
query     In the list of tokens, identify 'Person', 'Loc...
answer    It:O\nwill:O\nbe:O\neasy:O\nfor:O\nthem:O\nto:...
label     [O, O, O, O, O, O, O, O, O, O, O, O, O, O, O, ...
token     [It, will, be, easy, for, them, to, break, int...
Name: 366, dtype: object


# Áp dụng mô hình tiền huấn luyện (FinBERT / BERT + NER head)

In [65]:
from transformers import AutoTokenizer, AutoModelForTokenClassification, pipeline

tokenizer = AutoTokenizer.from_pretrained("dslim/bert-base-NER")
model = AutoModelForTokenClassification.from_pretrained("dslim/bert-base-NER")

ner_pipeline = pipeline(
    "ner",
    model=model,
    tokenizer=tokenizer,
    aggregation_strategy="simple",
    device=-1  # chạy trên CPU
)

sample_text = df_ner['text'].iloc[0]  # hoặc df_finer['text']
ner_pipeline(sample_text)


Some weights of the model checkpoint at dslim/bert-base-NER were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu


[{'entity_group': 'ORG',
  'score': np.float32(0.5220932),
  'word': '##K',
  'start': 85,
  'end': 86},
 {'entity_group': 'ORG',
  'score': np.float32(0.99310535),
  'word': 'Bank',
  'start': 90,
  'end': 94},
 {'entity_group': 'MISC',
  'score': np.float32(0.99309134),
  'word': 'California',
  'start': 101,
  'end': 111},
 {'entity_group': 'LOC',
  'score': np.float32(0.93892473),
  'word': 'Tasman Drive',
  'start': 174,
  'end': 186},
 {'entity_group': 'LOC',
  'score': np.float32(0.9982333),
  'word': 'Santa Clara',
  'start': 189,
  'end': 200},
 {'entity_group': 'LOC',
  'score': np.float32(0.99935454),
  'word': 'California',
  'start': 203,
  'end': 213},
 {'entity_group': 'LOC',
  'score': np.float32(0.9796879),
  'word': 'William St.',
  'start': 264,
  'end': 276},
 {'entity_group': 'LOC',
  'score': np.float32(0.9951899),
  'word': 'Ste',
  'start': 278,
  'end': 281}]

# Đánh giá mô hình (Evaluation)

In [69]:
import ast
import re

# Chuẩn hóa cột label
def fix_label_format(label_str):
    fixed = re.sub(r'(?<=\[|\s|,)(O|B-[A-Z]+|I-[A-Z]+)(?=,|\s|\])', r'"\1"', label_str)
    return ast.literal_eval(fixed)

df_ner['label_fixed'] = df_ner['label'].apply(fix_label_format)


In [70]:
df_ner['token_fixed'] = df_ner['text'].apply(lambda x: x.split())


In [73]:
from transformers import pipeline, AutoTokenizer, AutoModelForTokenClassification

tokenizer = AutoTokenizer.from_pretrained("dslim/bert-base-NER")
model = AutoModelForTokenClassification.from_pretrained("dslim/bert-base-NER")
ner_pipeline = pipeline("ner", model=model, tokenizer=tokenizer, aggregation_strategy="simple", device=-1)  # dùng CPU

y_true = df_ner['label_fixed'].tolist()
y_pred = []

for tokens in df_ner['token_fixed']:
    text = " ".join(tokens)
    result = ner_pipeline(text)

    pred_labels = ['O'] * len(tokens)
    for entity in result:
        word = entity['word']
        label = entity['entity_group']

        for i, token in enumerate(tokens):
            if word.lower() in token.lower():
                pred_labels[i] = 'B-' + label
                break

    y_pred.append(pred_labels)


Some weights of the model checkpoint at dslim/bert-base-NER were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu


In [74]:
from seqeval.metrics import classification_report

print(classification_report(y_true, y_pred))


              precision    recall  f1-score   support

         LOC       0.47      0.19      0.27       132
        MISC       0.00      0.00      0.00         0
         ORG       0.21      0.36      0.26       230
         PER       0.89      0.19      0.31       529

   micro avg       0.27      0.23      0.25       891
   macro avg       0.39      0.18      0.21       891
weighted avg       0.65      0.23      0.29       891



/usr/local/lib/python3.11/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [80]:
import re
import ast

# Chuẩn hóa nhãn và token
def fix_label_format(label_str):
    fixed = re.sub(r'(?<=\[|\s|,)(O|B-[A-Z]+|I-[A-Z]+)(?=,|\s|\])', r'"\1"', label_str)
    return ast.literal_eval(fixed)

def fix_token_format(token_str):
    token_str = token_str.replace("’", "'").replace("[", "").replace("]", "")
    raw_tokens = token_str.split(",")
    return [t.strip().strip('"').strip("'") for t in raw_tokens if t.strip()]

# Áp dụng vào DataFrame
df_finer['label_fixed'] = df_finer['label'].apply(fix_label_format)
df_finer['token_fixed'] = df_finer['token'].apply(fix_token_format)


In [81]:
from transformers import pipeline, AutoTokenizer, AutoModelForTokenClassification

tokenizer = AutoTokenizer.from_pretrained("dslim/bert-base-NER")
model = AutoModelForTokenClassification.from_pretrained("dslim/bert-base-NER")
ner_pipeline = pipeline("ner", model=model, tokenizer=tokenizer, aggregation_strategy="simple", device=-1)


Some weights of the model checkpoint at dslim/bert-base-NER were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu


In [83]:
y_true = df_finer['label_fixed'].tolist()
y_pred = []

for tokens in df_finer['token_fixed']:
    text = " ".join(tokens)
    result = ner_pipeline(text)

    pred_labels = []
    for token in tokens:
        label = 'O'
        for ent in result:
            if token.lower() in ent['word'].lower():
                label = 'B-' + ent['entity_group']
                break
        pred_labels.append(label)

    y_pred.append(pred_labels)


In [84]:
y_true_aligned = []
y_pred_aligned = []

for yt, yp in zip(y_true, y_pred):
    if len(yt) == len(yp):
        y_true_aligned.append(yt)
        y_pred_aligned.append(yp)

from seqeval.metrics import classification_report
print(classification_report(y_true_aligned, y_pred_aligned))


              precision    recall  f1-score   support

         LOC       0.30      0.49      0.37        77
        MISC       0.00      0.00      0.00         0
         ORG       0.23      0.50      0.31       170
         PER       0.15      0.28      0.20       118

   micro avg       0.18      0.43      0.25       365
   macro avg       0.17      0.32      0.22       365
weighted avg       0.22      0.43      0.29       365



/usr/local/lib/python3.11/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


#  Đánh giá kết quả NER trên hai tập dữ liệu FLARE

## 1. So sánh kết quả

| Loại dữ liệu     | Dataset `flare_ner` (ảnh 1) | Dataset `flare_finer_ord` (ảnh 2) |
|------------------|-----------------------------|------------------------------------|
| Tổng nhãn        | 891                         | 365                                |
| F1-score (micro) | **0.25**                    | **0.25**                           |
| F1-score (macro) | 0.21                        | **0.22**                           |
| Weighted F1      | 0.29                        | **0.29**                           |
| Nhãn mạnh nhất   | PER (precision: 0.89)       | ORG (recall: 0.50)                 |
| Nhãn yếu nhất    | MISC (không xuất hiện)      | MISC (không xuất hiện)             |

---

## 2. Nhận xét chi tiết

###  Dataset `flare_ner` (dạng câu):
- Mô hình cho precision **rất cao với nhãn PER** (`0.89`) nhưng recall thấp (`0.19`) → mô hình **chỉ đánh nhãn khi cực kỳ chắc chắn**.
- ORG và LOC có recall tương đối nhưng precision thấp → **khả năng phân biệt thực thể chưa ổn định**.
- Ngữ cảnh đầy đủ, nhưng dữ liệu chưa được token hóa rõ ràng nên khó đánh nhãn theo từ.

###  Dataset `flare_finer_ord` (dạng token):
- Precision toàn cục thấp hơn, nhưng recall cao hơn hẳn.
- **ORG (0.31 F1) và LOC (0.37 F1)** là hai nhãn được nhận diện tốt hơn.
- Dữ liệu rõ ràng hơn về định dạng token → phù hợp hơn với đánh giá hoặc fine-tuning mô hình.

---

## 3. Kết luận cuối cùng

- Mô hình `dslim/bert-base-NER` chỉ đạt khoảng **25% F1-score** → không đáp ứng yêu cầu cho trích xuất thực thể trong văn bản tài chính.
- Dữ liệu `flare_ner` có ngữ cảnh đầy đủ, thích hợp cho mô hình cần hiểu toàn câu.
- Dữ liệu `flare_finer_ord` gọn nhẹ, dễ huấn luyện, cho recall tốt hơn.
- Nhãn `MISC` bị bỏ qua hoàn toàn → cần làm giàu thêm nhãn này nếu huấn luyện mô hình mới.

---

